File: src/main/java/com/beishuo/service/AuthService.java
package com.beishuo.service;

import com.beishuo.common.ResultCode;
import com.beishuo.common.exception.BusinessException;
import com.beishuo.entity.User;
import com.beishuo.repository.UserRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;

@Service
public class AuthService {

    private static final Logger logger = LoggerFactory.getLogger(AuthService.class);

    @Autowired
    private UserRepository userRepository;

    @Autowired(required = false)
    private PasswordEncoder passwordEncoder;

    @Value("${cache.user-info-ttl:1800}")
    private Long userInfoTtl;

    /**
     * 用户注册
     */
    public Map<String, Object> register(String email, String password, String username) {
        // 检查邮箱是否已存在
        if (userRepository.existsByEmail(email)) {
            throw new BusinessException(ResultCode.USER_ALREADY_EXISTS);
        }

        // 确保用户名唯一，如果已存在则添加随机后缀
        String uniqueUsername = username;
        if (userRepository.existsByUsername(username)) {
            uniqueUsername = username + "_" + System.currentTimeMillis() % 10000;
        }

        // 创建用户
        User user = new User();
        user.setEmail(email);
        user.setUsername(uniqueUsername);
        user.setDisplayName(username);
        user.setPasswordHash(encodePassword(password));
        user.setRole("user");

        // 保存用户
        User savedUser = userRepository.save(user);
        
        logger.info("用户注册成功: email={}, username={}", email, uniqueUsername);
        return convertToMap(savedUser);
    }

    /**
     * 用户登录
     */
    public Map<String, Object> login(String email, String password) {
        // 查询用户
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new BusinessException(ResultCode.INVALID_CREDENTIALS));

        // 验证密码
        if (!verifyPassword(password, user.getPasswordHash())) {
            throw new BusinessException(ResultCode.INVALID_CREDENTIALS);
        }

        logger.info("用户登录成功: email={}", email);
        return convertToMap(user);
    }

    /**
     * 根据ID获取用户信息
     */
    public Map<String, Object> getUserById(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ResultCode.USER_NOT_FOUND));
        return convertToMap(user);
    }

    /**
     * 更新用户信息
     */
    public void updateUser(Long userId, Map<String, Object> userData) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new BusinessException(ResultCode.USER_NOT_FOUND));
        
        if (userData.containsKey("username")) {
            user.setUsername((String) userData.get("username"));
        }
        if (userData.containsKey("display_name")) {
            user.setDisplayName((String) userData.get("display_name"));
        }
        if (userData.containsKey("avatar_url")) {
            user.setAvatarUrl((String) userData.get("avatar_url"));
        }
        if (userData.containsKey("password")) {
            String password = (String) userData.get("password");
            user.setPasswordHash(encodePassword(password));
        }

        userRepository.save(user);
        logger.info("用户信息更新成功: userId={}", userId);
    }

    /**
     * 转换 User 实体为 Map
     */
    private Map<String, Object> convertToMap(User user) {
        Map<String, Object> map = new HashMap<>();
        map.put("id", user.getId());
        map.put("username", user.getUsername());
        map.put("email", user.getEmail());
        map.put("display_name", user.getDisplayName());
        map.put("avatar_url", user.getAvatarUrl());
        map.put("role", user.getRole());
        map.put("created_at", user.getCreatedAt());
        map.put("updated_at", user.getUpdatedAt());
        return map;
    }

    /**
     * 加密密码
     */
    private String encodePassword(String password) {
        if (passwordEncoder != null) {
            return passwordEncoder.encode(password);
        }
        // 如果没有配置PasswordEncoder，使用简单的哈希（生产环境必须配置）
        logger.warn("未配置PasswordEncoder，使用简单哈希（不安全）");
        return String.valueOf(password.hashCode());
    }

    /**
     * 验证密码
     */
    private boolean verifyPassword(String rawPassword, String encodedPassword) {
        if (passwordEncoder != null) {
            return passwordEncoder.matches(rawPassword, encodedPassword);
        }
        // 如果没有配置PasswordEncoder，使用简单比较（生产环境必须配置）
        logger.warn("未配置PasswordEncoder，使用简单比较（不安全）");
        return String.valueOf(rawPassword.hashCode()).equals(encodedPassword);
    }
}

File: python/app/services/auth_service.py
from typing import Optional, Dict, Any
from app.client.database_client import DatabaseClient
from app.client.redis_client import RedisClient
from app.core.exceptions import BusinessException
from app.common.result_code import ResultCode
from app.config import settings
from app.utils.password import PasswordUtil
from app.utils.logger import logger

class AuthService:
    """认证服务"""
    
    def __init__(self):
        self.database_client = DatabaseClient()
        self.redis_client = RedisClient()
        self.password_util = PasswordUtil()
    
    async def register(self, email: str, password: str, username: str) -> Dict[str, Any]:
        """用户注册"""
        # 检查用户是否已存在
        existing_user = await self.database_client.get_user_by_email(email)
        if existing_user:
            raise BusinessException(ResultCode.USER_ALREADY_EXISTS)
        
        # 加密密码
        hashed_password = self.password_util.hash_password(password)
        
        # 创建用户
        user_data = {
            "email": email,
            "password": hashed_password,
            "username": username
        }
        
        user = await self.database_client.create_user(user_data)
        if not user:
            raise BusinessException(ResultCode.INTERNAL_SERVER_ERROR, "用户创建失败")
        
        logger.info(f"用户注册成功: email={email}, username={username}")
        return user
    
    async def login(self, email: str, password: str) -> Dict[str, Any]:
        """用户登录"""
        # 查询用户
        user = await self.database_client.get_user_by_email(email)
        if not user:
            raise BusinessException(ResultCode.INVALID_CREDENTIALS)
        
        # 验证密码
        stored_password = user.get("password")
        if not self.password_util.verify_password(password, stored_password):
            raise BusinessException(ResultCode.INVALID_CREDENTIALS)
        
        # 缓存用户信息
        await self.redis_client.set(
            f"user:{user['id']}",
            user,
            timeout=settings.cache_user_info_ttl
        )
        
        logger.info(f"用户登录成功: email={email}")
        return user
    
    async def get_user_by_id(self, user_id: int) -> Dict[str, Any]:
        """根据ID获取用户信息"""
        # 先查缓存
        cached_user = await self.redis_client.get(f"user:{user_id}")
        if cached_user:
            return cached_user
        
        # 查数据库
        user = await self.database_client.get_user_by_id(user_id)
        if not user:
            raise BusinessException(ResultCode.USER_NOT_FOUND)
        
        # 缓存结果
        await self.redis_client.set(
            f"user:{user_id}",
            user,
            timeout=settings.cache_user_info_ttl
        )
        
        return user
    
    async def close(self):
        """关闭客户端连接"""
        await self.database_client.close()
        await self.redis_client.close()

